### ISSUE TITLE
Pipeline is unable to handle broken source links

### ISSUE THREAD

**Issue by KarenJewell** [OpenDataScotland/the_od_bods#193]
**Describe the bug**
Current [opendata.scot_pipeline](https://github.com/OpenDataScotland/opendata.scot_pipeline) can't cope when a source link is broken/offline. Current workaround is to manually remove the affected link from sources.csv before the pipeline runs otherwise it will fail.

**To Reproduce**
1. Add a fake or broken link to [sources.csv](https://github.com/OpenDataScotland/the_od_bods/blob/main/sources.csv)
2. Manually trigger [pipeline](https://github.com/OpenDataScotland/opendata.scot_pipeline/actions/workflows/pipeline.yml) to run
3. See error

**Expected behavior**
- [ ] The pipeline should successfully run to completion if any of the links are broken or otherwise invalid. 
- [ ] It should also inform which links are invalid
    - [ ] at minimum a print statement is fine, 
    - [ ] writing to a log file or 
    - [ ] posting to slack is also possible

**Screenshots**
None

**Hardware and software used**
 N/A

**Additional context**
Originally discussed on slack 21 Oct 2022: https://opendatascotland.slack.com/archives/C02HEHDL8AY/p1666345626928029

**Comment by andrewsage** [comment]
I thought I wrote something that checks for broken links back at CTC26? Or was that for something else?

**Comment by KarenJewell** [comment]
Yep you wrote the action which creates a github issue when a link goes down - which works magic, thanks! Because that's how we were alerted the dundee city site was down to begin with.

However, we do nothing with that knowledge and if left untreated, it cripples the pipeline, which is what this issue needs to resolve.

**Comment by andrewsage** [comment]
Thanks for the clarification. I didn't want any unnecessary wheel reinvention if it was the same thing.
Would a possible solution be to create a verified list that is passed into the pipeline instead of the raw list? Run a preprocess step to build the list?

**Comment by KarenJewell** [comment]
No I wouldn't want to duplicate work either - not when we have 62 other issues! 😬

That is definitely a shout, or a slight variation on it could be to extend your script to remove the broken link from source.csv if it is found to be broken on check, and add it back when restored.

So it would need a parent list to check and a child list to execute from in pipeline.

So a choice between modifying the sources list on the daily checks or executing a check pre-pipeline. 

I'm leaning towards the first as the second will add time on to pipeline run, and certainly more as that list gets longer.

**Comment by JackGilmore** [comment]
I still think an in-pipeline check or error handling is a good way to go. There still could be a chance that between doing your pre-pipeline check and then actually querying the portal that it could go down or an individual API endpoint could be broken but the rest of the API is fine. 

Deffo a good idea to log these failures as well to somewhere like slack to make them more visible to everyone.

**Comment by KarenJewell** [comment]
Do we know how long the current checks take?

**Comment by KarenJewell** [comment]
Do we know how long the current checks take?

Fair point about the links being broken between check and run, but that can happen anytime in execution as well or if anything hits an issue mid-run.

So I think extending the checks is great but that should not discount that we still need error handling regardless.

**Comment by JackGilmore** [comment]
Website status checker takes 30-50 seconds depending on how many links fail as it waits for timeout. You can see run stats here: https://github.com/OpenDataScotland/the_od_bods/actions/workflows/alive.yml

Our pipeline takes 5-10 mins to run with CKAN and NLS taking the longest time in that script as they usually have to do a web request per-dataset and nothing runs in parallel. It's on my mental "to do list" to see if we can run steps in the pipeline in parallel. Stats here: https://github.com/OpenDataScotland/opendata.scot_pipeline/actions/workflows/pipeline.yml

**Comment by JackGilmore** [comment]
I know we're trying to reduce the number of issues we have open, but it would be cool if we could graph the duration of time it takes for the pipeline to run each week

**Comment by KarenJewell** [comment]
That's great thanks, a minute extra isn't so much a problem, we can revisit if it does become an issue later.

Sounds to me the task list is then:
- [ ] Add sources alive checks to start of pipeline queue
- [ ] Create a child or temporary sources list in pipeline: list of **valid** links output from alive checks
- [ ] publish list of **invalid** links output from alive checks to slack 
- [ ] Change pipeline to execute from child/temporary **valid** sources list instead of current parent sources.csv
- [ ] Add error handling in pipeline if any item in iteration of sources fails (continue to next source)


Sound about right?

**Comment by KarenJewell** [comment]
On the back of PR #200, modified task list to be less convoluted:
- [ ] Add sources alive checks to start of pipeline queue
- [ ] Add error handling in pipeline if any item in iteration of sources fails (continue to next source)
- [ ] publish list of **invalid** links output to log and/or slack  

PR #200 address items 2 and 3 (publish to log). Will need a separate PR to JKAN repo for items 1 and 3 (post to slack)

**Comment by KarenJewell** [comment]
Re-opening issue as there are still outstanding pipeline changes which are github action changes not repo code changes so unable to tag to PR.

Item 2 above is complete with merged PR #200

**Comment by nutcracker22** [comment]
The log file is now working, as far as I can see. Is the source check at the begin of pipeline queue still required, since the check is done in [processor.py](https://github.com/OpenDataScotland/the_od_bods/blob/main/processor.py#L41)?

**Comment by JackGilmore** [comment]
Apologies, I've fallen behind with dealing with project admin recently so only just getting round to this.

@nutcracker22 Can you advise what you mean by the source check at the start of the pipeline? I'm not sure what you're referring to.

**Comment by nutcracker22** [comment]
@JackGilmore, I was just referring to [this comment](https://github.com/OpenDataScotland/the_od_bods/issues/193#issuecomment-1291247313), where Karen re-posted the list of open tasks. Since task 1 is not done at the start of the pipeline queue, but each link is checked (and logged, if not working), whenever processor.py accesses it, I wanted to know whether there's more to do on this task.

My goal is to resume working on this issue (including web scrapers in the error handling).

**Comment by JackGilmore** [comment]
Ah right. No, we probably don't need to do the source check at the beginning of the pipeline queue as there's appropriate error handling in place now whenever we're requesting JSON from data portal APIs.

Do you think we could close this now? @KarenJewell

**Comment by KarenJewell** [comment]
If we're saying 2 and 3 is done, and 1 is redundant then sure let's close it.
